crates/re_viewer/build.rs

```
@@ -1,4 +1,233 @@
+}
+
+macro_rules! bail {
+    ($lit:literal) => (return Err(error!($lit)));
+    ($($tt:tt)*) => (return Err(error!($($tt)*).into()));
+}
+
+fn git_branch_name(sh: &Shell) -> Result<String> {
+    Ok(String::from_utf8(
+        cmd!(sh, "git rev-parse --abbrev-ref HEAD").output()?.stdout,
+    )?)
+}
+
+fn git_short_hash(sh: &Shell) -> Result<String> {
+    let full_hash = String::from_utf8(cmd!(sh, "git rev-parse HEAD").output()?.stdout)?;
+    Ok(full_hash.trim()[0..7].to_string())
+}
+
+fn parse_release_version(branch: &str) -> Option<String> {
+    // release-\d+.\d+.\d+
+
+    let branch = branch.strip_prefix("release-")?;
+
+    let mut version_parts = branch.split('.');
+    let major = version_parts.next()?.parse::<u8>().ok()?;
+    let minor = version_parts.next()?.parse::<u8>().ok()?;
+    let patch = version_parts.next()?.parse::<u8>().ok()?;
+
+    if version_parts.next().is_some() {
+        return None;
+    }
+
+    Some(format!("{major}.{minor}.{patch}"))
+}
+
+#[derive(serde::Deserialize)]
+struct Frontmatter {
+    title: Option<String>,
+    tags: Option<Vec<String>>,
+    description: Option<String>,
+    thumbnail: Option<String>,
+    thumbnail_dimensions: Option<[u64; 2]>,
+    #[serde(default)]
+    demo: bool,
+}
+
+#[derive(serde::Serialize)]
+struct ManifestEntry {
+    name: String,
+    title: String,
+    description: String,
+    tags: Vec<String>,
+    demo_url: String,
+    rrd_url: String,
+    thumbnail: Thumbnail,
+}
+
+impl ManifestEntry {
+    fn new(example: Example, base_url: &str) -> Result<Self> {
+        macro_rules! get {
+            ($e:ident, $f:ident) => {
+                match $e.readme.$f {
+                    Some(value) => value,
+                    None => bail!("{:?} is missing field {:?}", $e.name, stringify!($f)),
+                }
+            };
+        }
+
+        let thumbnail_dimensions = get!(example, thumbnail_dimensions);
+
+        Ok(Self {
+            title: get!(example, title),
```

Why use a macro to unwrap the `title: Option<String>` instead of using `title: String` directly?

---

I think I just wanted the parsing from YAML to not fail early, so that I could check for the presence of `build_args`. I just defaulted all the fields instead.